utils/datasets/cifar10_dataset.py

```diff
@@ -0,0 +1,330 @@
+# Copyright 2019, Google LLC.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""Library for loading and preprocessing CIFAR-10 training and testing data."""
+
+import collections
+from typing import Tuple
+import numpy as np
+import tensorflow as tf
+import tensorflow_federated as tff
+from tensorflow.keras import datasets, layers, models
+
+CIFAR_SHAPE = (32, 32, 3)
+TOTAL_FEATURE_SIZE = 32 * 32 * 3
+NUM_EXAMPLES_PER_CLIENT = 5000
+
+def load_cifar10_federated(
+    num_clients: int = 10, 
+    num_classes: int = 10, 
+    alpha: float = 1, 
+    train_client_batch_size: int = 20, 
+    test_client_batch_size: int = 100):
+  '''Construct a synthetic federated CIFAR-10 from its centralized version
+
+  Loads the train dataset into a non iid distribution over clients using the
+  sampling method based on LDA, taken from this paper:
+  Measuring the Effects of Non-Identical Data Distribution for
+  Federated Visual Classification (https://arxiv.org/pdf/1909.06335.pdf).
+
+  Args:
+    num_clients: An integer specifing the total number of clients.
+    num_classes: An integer representing the number of lables in the dataset.
+    alpha: A float controling the data heterogeneity among clients.
+    train_client_batch_size: A float representing the batch size during 
+      training.
+    test_client_batch_size: A float representing the batch size during
+      test.
+
+  Returns:
+    A tuple of `tff.simulation.ClientData` representing unpreprocessed
+    train data and test data.
+  '''
+  (train_images, train_labels), (test_images, test_labels) \
+      = datasets.cifar10.load_data()
+  client_train = collections.OrderedDict()
+  client_test = collections.OrderedDict()
+
+  # The following structure will store the image index for each client
+  idx_batch = [[] for _ in range(num_clients)]
+
+  for k in range(num_classes):
+    # For each class(label) sample in a proportion for each client. 
+    # The proportion is determined by a dirichlet distribution.
+    label_k = np.where(train_labels.squeeze()==k)[0]
+    np.random.shuffle(label_k)
+    proportion = np.random.dirichlet(alpha*np.ones((num_clients,)))
```

Comment:
This is a different approach from Hsu et al., 2019. This partitions each label space according to some categorical distribution over clients, instead of making each client have a categorical distribution over labels.

I think the end result is similar, but I don't think it's identical. One important distinction: This approach can create massively unbalanced clients. Is there a specific reason to do the sampling in this manner?

Reply:
For posterity: Appendix B of https://arxiv.org/pdf/2003.08082.pdf describes how to do the LDA client partitioning in a balanced way, when sampling examples without replacement. I think we should be doing the partitioning in that way.